source/interaction/interaction_utilities.cc

```diff
@@ -1132,6 +1133,387 @@ namespace fdl
+                          t = (p - r(1)) / q[1];
+                        }
+                      t_vals.push_back(std::make_pair(t, Point<1>(u)));
+                      std::cout << "line parameter t:"<<t<< std::endl;
+
+                    }
+                }
+            };
+            if(mapping.get_degree()==2){
+                const dealii::Point<2> p0 = mapping.transform_unit_to_real_cell(elem, Point<1>(0.0));
+                const dealii::Point<2> p1 = mapping.transform_unit_to_real_cell(elem,Point<1>(1));
+                const dealii::Point<2> p2 = mapping.transform_unit_to_real_cell(elem,Point<1>(0.5));
+                intersect_line_with_Q2_edge ( t_vals,p0,p1, p2, r, q,tol);}
+            else{
+                for (unsigned int j = 0; j <(pow(2,d_intersection_refinement)); ++j){
+                    const dealii::Point<2> p0 = mapping.transform_unit_to_real_cell(elem, Point<1>(1.0/(pow(2,d_intersection_refinement))*j));
+                    const dealii::Point<2> p1 = mapping.transform_unit_to_real_cell(elem,Point<1>(1.0/(pow(2,d_intersection_refinement))*(j+1)));
+                    const dealii::Point<2> p2 = mapping.transform_unit_to_real_cell(elem,Point<1>(1.0/(pow(2,d_intersection_refinement))*(j+0.5)));
+                    intersect_line_with_Q2_edge ( t_vals,p0,p1, p2, r, q,tol);}
+              }
+
+          }
+      }
+
+    return is_interior_intersection;
+  } // intersect_line_with_edge
+
+  //WARNING: This code is specialized to the case in which q is a unit vector
+  //aligned with the coordinate axes.
+  bool
+  intersect_line_with_face(std::vector<std::pair<double, Point<2>> >& t_vals,
+                           const typename DoFHandler<2, 3>::active_cell_iterator elem,
+                           const dealii::MappingQ<2, 3> &mapping,
+                           dealii::Point<3> r,
+                           dealii::Tensor<1,3> q,
+                           const double tol)
+  {
+    bool is_interior_intersection = false;
+    t_vals.resize(0);
+    elem->get_triangulation().get_manifold(1);
+    int d_intersection_refinement =5;
+    switch (mapping.get_degree())
+      {
+        case 1:
+          {
+            const dealii::Point p00 = elem->vertex(0);
+            const dealii::Point p10 = elem->vertex(1);
+            const dealii::Point p01 = elem->vertex(2);
+            const dealii::Point p11 = elem->vertex(3);
+
+            const dealii::Tensor<1,3> a = p11 - p10 - p01 + p00;
+            const dealii::Tensor<1,3> b = p10 - p00;
+            const dealii::Tensor<1,3> c = p01 - p00;
+            const dealii::Tensor<1,3> d = p00;
+
+            double A1, A2, B1, B2, C1, C2, D1, D2;
+            if (q[0] != 0.0)
+              {
+                A1 = a[1];
+                A2 = a[2];
+                B1 = b[1];
+                B2 = b[2];
+                C1 = c[1];
+                C2 = c[2];
+                D1 = d[1] - r(1);
+                D2 = d[2] - r(2);
```

Comment:
Can we rewrite this using `Tensor`? This code is excessively difficult to understand.